extension/src/frequency.rs

```diff
@@ -403,100 +400,212 @@ pub mod toolkit_experimental {
+                        Some(100) // TODO: default OID, there should be a constant for this
+                    } else {
+                        get_collation(fcinfo)
+                    };
+                    SpaceSavingTransState::freq_agg_from_type_id(freq, typ, collation).into()
+                },
+                Some(state) => state,
+            };
+    
+            state.add(value.into());
+            Some(state)
+        })
     }
+}
 
-    fn finally(
-        state: Option<&mut State>,
-    ) -> Option<toolkit_experimental::FrequencyAggregate<'static>> {
-        state.map(FrequencyAggregate::from)
+#[pg_extern(schema = "toolkit_experimental", immutable, parallel_safe)]
+pub fn space_saving_combine(
+    state1: Internal,
+    state2: Internal,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Internal {
+    unsafe {
+        space_saving_combine_inner(state1.to_inner(), state2.to_inner(), fcinfo).internal()
     }
-
-    fn serialize(state: &State) -> bytea {
-        crate::do_serialize!(state)
+}
+pub fn space_saving_combine_inner(
+    a: Option<Inner<SpaceSavingTransState>>,
+    b: Option<Inner<SpaceSavingTransState>>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Inner<SpaceSavingTransState>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            match (a, b) {
+                (Some(a), Some(b)) => Some(SpaceSavingTransState::combine(&*a, &*b).into()),
+                (Some(a), None) => Some(a.clone().into()),
+                (None, Some(b)) => Some(b.clone().into()),
+                (None, None) => None,
+            }
+        })
     }
+}
 
-    fn deserialize(bytes: crate::raw::bytea) -> State {
-        crate::do_deserialize!(bytes, FrequencyTransState)
-    }
+#[pg_extern(schema = "toolkit_experimental", immutable, parallel_safe)]
+fn space_saving_final(
+    state: Internal,
+    _fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<toolkit_experimental::SpaceSavingAggregate<'static>> {
+    let state: Option<&SpaceSavingTransState> = unsafe { state.get() };
+    state.map(SpaceSavingAggregate::from)
+}
 
-    fn combine(a: Option<&State>, b: Option<&State>) -> Option<State> {
-        match (a, b) {
-            (Some(a), Some(b)) => Some(FrequencyTransState::combine(a, b)),
-            (Some(a), None) => Some(a.clone()),
-            (None, Some(b)) => Some(b.clone()),
-            (None, None) => None,
-        }
-    }
+#[pg_extern(schema = "toolkit_experimental", immutable, parallel_safe)]
+fn space_saving_serialize(
+    state: Internal,
+) -> bytea {
+    let state: Inner<SpaceSavingTransState> = unsafe { state.to_inner().unwrap() };
+    crate::do_serialize!(state)
+}
+
+#[pg_extern(schema = "toolkit_experimental", immutable, parallel_safe)]
+pub fn space_saving_deserialize(
+    bytes: bytea,
+    _internal: Internal,
+) -> Internal {
+    let i: SpaceSavingTransState = crate::do_deserialize!(bytes, SpaceSavingTransState);
+    Inner::from(i).internal()
 }
 
+extension_sql!("\n\
```

Comment:
After reading this whole change, I didn't see why we couldn't stick with the aggregate builder. So I went ahead and tried restoring it, and it did work. I also implemented my suggestion above to factor out a transition function to share among the three aggregates.

The only down side is one of the three aggregates cannot use the builder as the builder has no support for giving two aggregates the same name. It should not be too hard to teach it to do that, but I'm not yet expert enough in macros to do that. I'd like to look into that soon.

But I think it reads quite a bit better with two aggregate builders and one `CREATE AGGREGATE`.

I'll push a branch with what I did.

Reply:
I'm actually not a fan of this approach for a few reasons.
1. It duplicates all the support functions (serialize, combine, etc.), which is clunky and error prone.  It's even worse in this case as all of these duplicated functions will get corresponding postgres functions created for them.
2. The `CREATE AGGREGATE` call is using the (ugly named) functions implicitly generated by the macro.  This makes the code almost impossible to follow unless you're familiar with the internals of the macro.
3. The `topn_agg_with_skew_trans` exposes all of the messy details that we're trying to hide with the macro.
4. I do like combining the bulk of the transition functions, but the functor code reads a little odd to me.  This is probably just my rust inexperience though.

I realize that a couple of these points could be handled by fixing the macro to allow name reuse, though this would also make the first point worse (I'm also not sure how trivial this change really is).  However without this, I think mixing the aggregate builder and `CREATE AGGREGATE` approaches is almost the worst of both worlds.

At this point I'm inclined to stick with the multiple `CREATE AGGREGATE` calls to maximize the reuse of the functions.  I will refactor the transition functions though to be something more along the lines of what you're doing in your example.